compiler/compile/OMRCompilation.cpp

```diff
@@ -440,21 +440,25 @@ OMR::Compilation::Compilation(
       _options->setOption(TR_EnableOSROnGuardFailure, false);
       }
 
-   if (_options->getOption(TR_EnableHCR))
+   // Do not default OSR and HCR on if:
+   // NextGenHCR is disabled, as they are enabled for it
+   // OSR is explicitly disabled
+   // FSD is enabled, as HCR cannot be enabled with it
+   //
+   // Defaults are performed this late to account for the disable OSR above, as we do not want to enable HCR in cases where
+   // its not necessary
+   if (!self()->getOption(TR_DisableNextGenHCR) && !self()->getOption(TR_DisableOSR) && !self()->getOption(TR_FullSpeedDebug))
       {
-      if (!self()->getOption(TR_DisableOSR))
-         self()->setOption(TR_EnableOSR); // Make OSR the default for HCR
-      if (_options->getOption(TR_EnableNextGenHCR))
-         {
-         self()->setOption(TR_EnableOSR);
-         _options->setOption(TR_DisableOSR, false);
-         }
+      self()->setOption(TR_EnableOSR); // OSR and HCR must be enabled for NextGenHCR
+      self()->setOption(TR_EnableHCR);
       }
 
    if (_options->getOption(TR_EnableOSR))
       {
       // Current implementation of partial inlining will break OSR
       self()->setOption(TR_DisablePartialInlining);
+      // Can't yet transition a MH thunk frame into equivalent interpreter frames
+      self()->setOption(TR_DisableMethodHandleThunks);
```

Comment:
This has been addressed by disabling OSR when compiling MH thunks, rather than disabling MH thunks in OSR.